annotated.go

```diff
@@ -272,210 +529,357 @@ func (la *lifecycleHookAnnotation) apply(ann *annotated) error {
+
+	origFn := reflect.ValueOf(ann.Target)
+	newFnType := reflect.FuncOf(paramTypes, resultTypes, false)
+	newFn := reflect.MakeFunc(newFnType, func(args []reflect.Value) []reflect.Value {
+		origArgs := make([]reflect.Value, len(args))
+		copy(origArgs, args)
+		args = remapParams(args)
+		results := origFn.Call(args)
+		if hasError {
+			errVal := results[len(results)-1]
+			results = results[:len(results)-1]
+			if err, _ := errVal.Interface().(error); err != nil {
+				// if constructor returned error, do not call hook installer
+				return append(results, errVal)
+			}
+		}
+		hookInstallerResults := hookInstaller.Call(append(results, origArgs...))
+		results = append(results, hookInstallerResults[0])
+		return results
+	})
+	return newFn.Interface(), nil
+}
+
 var (
 	_typeOfLifecycle reflect.Type = reflect.TypeOf((*Lifecycle)(nil)).Elem()
 	_typeOfContext   reflect.Type = reflect.TypeOf((*context.Context)(nil)).Elem()
 )
 
-type valueResolver func(reflect.Value, int) reflect.Value
-
-func (la *lifecycleHookAnnotation) resolveMap(results []reflect.Type) (
-	resultMap map[reflect.Type]valueResolver,
+// buildHookInstaller returns a function that appends a hook to Lifecycle when called,
+// along with the new paramter types and a function that maps arguments to the annotated constructor
+func (la *lifecycleHookAnnotation) buildHookInstaller(ann *annotated) (
+	hookInstaller reflect.Value,
+	paramTypes []reflect.Type,
+	remapParams func([]reflect.Value) []reflect.Value, // function to remap parameters to function being annotated
 ) {
-	// index the constructor results by type and position to allow
-	// for us to omit these from the in types that must be injected,
-	// and to allow us to interleave constructor results
-	// into our hook arguments.
-	resultMap = make(map[reflect.Type]valueResolver, len(results))
-
-	for _, r := range results {
-		resultMap[r] = func(v reflect.Value, pos int) (value reflect.Value) {
-			return v
+	paramTypes = ann.currentParamTypes()
+	paramTypes, remapParams = injectLifecycle(paramTypes)
+
+	resultTypes, hasError := ann.currentResultTypes()
+	if hasError {
+		resultTypes = resultTypes[:len(resultTypes)-1]
+	}
+
+	ctxPos := -1
+	ctxStructPos := -1
+	origHookFn := reflect.ValueOf(la.Target)
+	origHookFnT := reflect.TypeOf(la.Target)
+	invokeParamTypes := []reflect.Type{
+		_typeOfLifecycle,
+	}
+	for i := 0; i < origHookFnT.NumIn(); i++ {
+		t := origHookFnT.In(i)
+		if t == _typeOfContext && ctxPos < 0 {
+			ctxPos = i
+			continue
+		}
+		if isIn(t) {
+			fields := []reflect.StructField{_inAnnotationField}
+			for j := 1; j < t.NumField(); j++ {
+				field := t.Field(j)
+				if field.Type == _typeOfContext && ctxPos < 0 {
+					ctxStructPos = i
+					ctxPos = j
+					continue
+				}
+				fields = append(fields, field)
+			}
+			invokeParamTypes = append(invokeParamTypes, reflect.StructOf(fields))
+			continue
 		}
+		invokeParamTypes = append(invokeParamTypes, origHookFnT.In(i))
```

Comment:
Is there any reason why this isn't just an if / else if / else?

Reply:
Or, what might be better is:

```
if t == _typeOfContext && ctxPos < 0 {
    ctxPos = i
    continue
if !isIn(t) {
    invokeParamTypes = append(invokeParamTypes, origHookFnT.in(i))
    continue
}
fields := []reflect.StructField{_inAnnotationField}
<...>
```

Reply:
Done